src/v/cloud_storage/materialized_resources.cc

```diff
@@ -456,4 +478,60 @@ void materialized_resources::maybe_trim_segment(
     }
 }
 
+class throttled_dl_source : public ss::data_source_impl {
+public:
+    explicit throttled_dl_source(
+      ss::input_stream<char> st,
```

Comment:
you can also called `input_stream<T>::detatch()` and wrap the underlying `data_source_impl` instead of the input stream, if that makes sense. does input stream have any state that we'd want to preserve after wrapping?